lightning/src/offers/offer.rs

```diff
@@ -0,0 +1,762 @@
+		self.offer.paths.get_or_insert_with(Vec::new).push(path);
+		self
+	}
+
+	/// Sets a fixed quantity of items for the offer. If not set, `1` is assumed.
+	///
+	/// Successive calls to this method or [`quantity_range`] will override the previous setting.
+	///
+	/// [`quantity_range`]: Self::quantity_range
+	pub fn quantity_fixed(mut self, quantity: NonZeroU64) -> Self {
+		let quantity = Some(quantity.get()).filter(|quantity| *quantity != 1);
+		self.offer.quantity_min = quantity;
+		self.offer.quantity_max = quantity;
+		self
+	}
+
+	/// Sets a quantity range of items for the offer. If not set, `1` is assumed.
+	///
+	/// Successive calls to this method or [`quantity_fixed`] will override the previous setting.
+	///
+	/// [`quantity_fixed`]: Self::quantity_fixed
+	pub fn quantity_range<R: RangeBounds<NonZeroU64>>(mut self, quantity: R) -> Self {
+		self.offer.quantity_min = match quantity.start_bound() {
+			Bound::Included(n) => Some(n.get()),
+			Bound::Excluded(_) => unreachable!(),
+			Bound::Unbounded => Some(1),
+		};
+		self.offer.quantity_max = match quantity.end_bound() {
+			Bound::Included(n) => Some(n.get()),
+			Bound::Excluded(n) => Some(n.get() - 1),
+			Bound::Unbounded => None,
+		};
+
+		// Use a minimal encoding whenever 1 can be inferred.
+		if let Some(1) = self.offer.quantity_min {
+			match self.offer.quantity_max {
+				Some(1) => {
+					self.offer.quantity_min = None;
+					self.offer.quantity_max = None;
+				},
+				Some(_) => {
+					self.offer.quantity_min = None;
+				},
+				None => {},
+			}
+		}
+
+		// Assume quantity isn't supported if the range is empty.
+		if self.offer.quantity_min() > self.offer.quantity_max() {
+			self.offer.quantity_min = None;
+			self.offer.quantity_max = None;
+		}
+
+		self
+	}
+
+	/// Builds an [`Offer`] from the builder's settings.
+	pub fn build(self) -> Offer {
+		let mut bytes = Vec::new();
+		self.offer.write(&mut bytes).unwrap();
+
+		Offer {
+			bytes,
+			contents: self.offer,
+		}
+	}
+}
+
+/// An `Offer` is a potentially long-lived proposal for payment of a good or service.
+///
+/// An offer is precursor to an `InvoiceRequest`. A merchant publishes an offer from which a
+/// customer may request an `Invoice` for a specific quantity and using a specific amount. Offers
+/// may be denominated in currency other than bitcoin but are ultimately paid using the latter.
+///
+/// Through the use of [`BlindedPath`]s, offers provide recipient privacy.
+#[derive(Clone, Debug)]
+pub struct Offer {
+	// The serialized offer. Needed when creating an `InvoiceRequest` if the offer contains unknown
+	// fields.
+	bytes: Vec<u8>,
+	contents: OfferContents,
+}
+
+/// The contents of an [`Offer`], which may be shared with an `InvoiceRequest` or an `Invoice`.
+#[derive(Clone, Debug)]
+pub(crate) struct OfferContents {
+	chains: Option<Vec<ChainHash>>,
+	metadata: Option<Vec<u8>>,
+	amount: Option<Amount>,
+	description: String,
+	features: Option<OfferFeatures>,
```

Comment:
> Could you clarify how that makes the API inconsistent? Seems a bit cleaner to avoid returning an `Option` where it makes sense.

There's a few questions that we'd have to answer. If we return `&OfferFeatures` instead of `Option<&OfferFeature>`, should we return
- `&str` or `Option<&String>`?
- `&[T]` or `Option<&Vec<T>>`
- `u64` or `Option<&Amount>`
- `Duration` or `Option<Duration>`

I can see the argument for `&[T]`, but I'm not as sure about `&str`. Is it better for the user to check against `None` or the empty string to determine if a string is set. Is it ok to treat the empty string as not set, and vice versa?

For `amount`, we'll eventually need an exchange rate interface, so this could work for each variant. But for `None`, should we return `0`? That said, presumably a wallet would want to display the offer's native currency, so would likely need to stick with `Option` or add a third variant to `Amount` for `None`.

Should we stick with `Option` for `absolute expiry` or return the maximum duration if `None`?

> I thought the `OfferTlvStreamRef` could work around the serialization issues but maybe I'm missing something

Yeah, I suppose it could by doing:

```rust
let features = if self.features == OfferFeatures::empty() { None } else { Some(&self.features) };
```

But that raises a secondary concern. When should `Option` be used in `OfferContents`? In some places like `amount` it probably should be. But what about `issuer` or `paths`? We'd have to be comfortable with treating an empty TLV value as the same as if the TLV record had not been set. This also means an encoded TLV might not be equal with it after parsed and re-encoded because presumably we wouldn't write the record for empty values. Similarly for `features`.

My intuition is to stick with `Option` whenever possible, especially if `InvoiceRequest` or `Invoice` need to check against the presence. But we probably can go without `Option` around `Vec` and `OfferFeatures` with the above caveat in mind, assuming we're ok with additional conversion logic in `as_tlv_stream` here and `try_from` in later PRs. I'm non sure about `String`, regardless of what the API ultimately looks like.

Let me know what you think.

Reply:
Ah, since we already don’t return `Option`s for all getters I didn’t think of that as much of an API inconsistency(?).

> * u64 or Option<&Amount>
> * Duration or Option<Duration>

I thought these two could make sense as options still. Certainly expiry being `None` makes sense IIUC? Fair point that amount could be weird though, is it possible to make `Amount`s contain `NonZeroU64`s? 

> &str or Option<&String>?

IIUC that’s only for `issuer`, IMO that one arguably makes sense as an `Option<&String>` for the reasons you stated, no strong feelings though.

`features` was the only one where I couldn’t see a reason for it to be `None` vs `Some(OfferFeatures::empty()` and therefore seemed weird in the public API. 

We could get rid of the `features()` getter entirely and only have individual methods for individual features (`offer.supports_mpp()`). That way it could stay an `Option` under the hood so that parsing-and-reencoding always gets the same result. I admit I hadn’t considered that situation though. 

Although, I thought since we save the offer `bytes`, parsing and re-encoding will always be the same anyway? Lmk what I'm missing, not sure I understood your point about how sticking with `Option` benefits `InvReq`/`Inv`

Reply:
> Ah, since we already don’t return `Option`s for all getters I didn’t think of that as much of an API inconsistency(?).
> 

Yes, very true.

> > * u64 or Option<&Amount>
> > * Duration or Option
> 
> I thought these two could make sense as options still. Certainly expiry being `None` makes sense IIUC? Fair point that amount could be weird though, is it possible to make `Amount`s contain `NonZeroU64`s?
> 

Hadn't thought about `NonZeroU64` for amount. Honestly, it's a little tough to work with, and the BOLT doesn't seem to care if it is zero. Just says it's a "minimum".

> > &str or Option<&String>?
> 
> IIUC that’s only for `issuer`, IMO that one arguably makes sense as an `Option<&String>` for the reasons you stated, no strong feelings though.
> 

Yeah, just wasn't sure if we should try to do something similar to `Vec`s. I think I'm learning towards `Option` but `Option<&str>`, which removes some boilerplate from tests.

> `features` was the only one where I couldn’t see a reason for it to be `None` vs `Some(OfferFeatures::empty()` and therefore seemed weird in the public API.
> 
> We could get rid of the `features()` getter entirely and only have individual methods for individual features (`offer.supports_mpp()`). That way it could stay an `Option` under the hood so that parsing-and-reencoding always gets the same result. I admit I hadn’t considered that situation though.
> 

Let's use `OfferFeatures::empty()` without the `Option` for now. Could be nice to know if there are odd, unknown features.

> Although, I thought since we save the offer `bytes`, parsing and re-encoding will always be the same anyway? Lmk what I'm missing, not sure I understood your point about how sticking with `Option` benefits `InvReq`/`Inv`

Ah, right! Literally ran into this yesterday when adding parsing failure tests in the next PR. Had to manually modify the TLV stream to make the test fail. 😛

Reply:
> Lmk what I'm missing, not sure I understood your point about how sticking with `Option` benefits `InvReq`/`Inv`

Sometime we need to check if a field in the `Offer` is present to determine how to handle the `InvoiceRequest` (e.g., `amount`). During parsing, it's simplest to convert the `OfferTlvStream` to `OfferContents` first before performing these checks. So having the `Option` in `OfferContents` can be useful.